File: runtime/app/schemas/intake.py
"""Pydantic contracts for the guided intake wizard (Phase CO-1A).

Mirrors packages/shared/src/intake.ts one-for-one (runtime is the contract owner).
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

# There is NO canonical step order any more (doc 40 §A4): the Intake Planner picks the next
# screen from app.intake.planner.SCREEN_REGISTRY after every capture. `current_step` on the
# wire is the planner's pick — a registry screen id, or "READY".


class ConfirmationPrompt(BaseModel):
    """A low-confidence extracted field for a trivial yes/no confirmation (P1)."""

    field: str
    read_value: str
    prompt: str
    confidence: float


class CapturedData(BaseModel):
    coverage: dict[str, Any] = Field(default_factory=dict)
    bills_count: int = 0
    eobs_count: int = 0
    visit_context: str | None = None
    # Sprint B (DL-82): the case's detected/confirmed coverage regime + the detection
    # metadata (candidate/confidence/evidence/verified) the confirm screen preselects
    # from. coverage_regime is null until high-confidence detection or a user confirm.
    coverage_regime: str | None = None
    regime_detection: dict[str, Any] | None = None


class PlanProposal(BaseModel):
    """A stored plan-level benefit design proposed for one-tap confirmation (CO-12C).
    Plan-level only — no PHI. The user confirms (writes through to coverage) or
    rejects (forks a corrected design)."""

    plan_library_id: str
    payer: str
    plan_name: str | None = None
    plan_year: int
    benefit_design: dict[str, Any] = Field(default_factory=dict)
    confidence: int = 1
    summary: str


class IntakeProgress(BaseModel):
    """The segmented bar (§A8). It never regresses: `segments[].filled` includes the persisted
    high-water mark, and `note` explains a segment a reclassified document no longer backs."""

    segments: list[dict[str, Any]] = Field(default_factory=list)
    filled: int = 0
    total: int = 7
    line: str | None = None  # "1 of 7 — nice start"; never a bare "Step N of M"
    note: str | None = None
    glosses: dict[str, Any] = Field(default_factory=dict)
    high_water: list[str] = Field(default_factory=list)


class IntakeResume(BaseModel):
    """"Pick up where you left off" (§C7) — shown when the user returns to /intake with an
    unfinished guided case. `link_expiry` states the REAL magic-link lifetime."""

    case_file_id: str
    title: str | None = None
    body: str | None = None
    primary: str | None = None
    new: str | None = None
    link_expiry: str | None = None


class IntakeStateResponse(BaseModel):
    # None until the user starts (POST /intake/start): opening the landing creates nothing.
    case_file_id: str | None = None
    intake_status: str  # not_started | in_progress | complete
    intake_mode: str = "guided"
    current_step: str  # the planner's pick: a SCREEN_REGISTRY id, or "READY"
    completed_steps: list[str] = Field(default_factory=list)  # filled progress groups
    # The screen to draw: id, kind, copy (registry strings), data, example?, help?, skippable.
    screen: dict[str, Any] = Field(default_factory=dict)
    progress: IntakeProgress = Field(default_factory=IntakeProgress)
    chrome: dict[str, str] = Field(default_factory=dict)
    resume: IntakeResume | None = None
    captured_data: CapturedData
    missing_items: list[str] = Field(default_factory=list)
    # CO-12C: a pending PlanLibrary proposal (rendered by the plan_rules_confirm screen).
    plan_proposal: PlanProposal | None = None


class StepAck(IntakeStateResponse):
    """Returned by every intake write. It IS the next state — the client never decides where
    to go, it draws `screen`. `confirmations` carries low-confidence card fields (P1)."""

    confirmations: list[ConfirmationPrompt] = Field(default_factory=list)


class IntakeAnswerRequest(BaseModel):
    """One answer from one screen. `action`: continue | skip | ack | yes | no | not_sure | fix.
    `values` is screen-specific and validated by the route against that screen's fields."""

    case_file_id: str
    screen: str
    action: str = "continue"
    values: dict[str, Any] = Field(default_factory=dict)


class IntakeRunResponse(BaseModel):
    case_file_id: str
    status: str
    # Where the app goes next — the EXISTING reveal/thread; the guided route builds no results UI.
    next_route: str
    conversation_id: str | None = None


class IntakeHandoffRequest(BaseModel):
    case_file_id: str


class HelpEmailRequest(BaseModel):
    case_file_id: str | None = None
    document_type: str
    screen: str | None = None


class VisitContextRequest(BaseModel):
    case_file_id: str | None = None
    visit_context: str = Field(max_length=500)


class ExtractRequest(BaseModel):
    case_file_id: str | None = None
    document_id: str


class RegimeConfirmRequest(BaseModel):
    """The user's explicit answer to 'How are you covered?' (Sprint B ladder). The
    route validates coverage_regime against the seven regimes and marks the detection
    verified=true (user_declared)."""

    case_file_id: str | None = None
    coverage_regime: str


class CompletionSummary(BaseModel):
    case_file_id: str
    intake_status: str
    captured: list[str] = Field(default_factory=list)
    missing_items: list[str] = Field(default_factory=list)
    summary: str

File: packages/shared/src/intake.ts
/**
 * Guided intake wizard API contracts (Phase CO-1A).
 *
 * Mirrors runtime/app/schemas/intake.py one-for-one. The runtime is the contract
 * owner; this is the typed mirror for the Expo wizard.
 */

// There is NO step list any more (doc 40 §A4). CO-1A's INTAKE_STEPS / IntakeStep /
// SKIPPABLE_STEPS are gone: the server's Intake Planner picks the next screen after every
// capture, and the client draws the `screen` it is handed. `current_step` is that pick — a
// screen id from the runtime's SCREEN_REGISTRY, or 'READY'.
export type IntakeStatus = 'not_started' | 'in_progress' | 'complete';

/** The renderer the client uses for a screen. A closed set — an unknown kind renders the
 *  generic info layout, so a server ahead of the app degrades instead of crashing. */
export type IntakeScreenKind =
  | 'info'
  | 'handoff'
  | 'capture'
  | 'coach'
  | 'summary'
  | 'choice'
  | 'fields'
  | 'plan_confirm'
  | 'timeline'
  | 'attest'
  | 'progress'
  | 'confirmations'
  | 'readiness'
  | 'ready';

export type IntakeProgressGroup =
  | 'bill'
  | 'card'
  | 'plan_rules'
  | 'eob'
  | 'timeline'
  | 'about_you'
  | 'confirmations';

export interface IntakeExample {
  ask: string;
  title: string | null;
  /**
   * doc 41 (Brock 2026-09-21, decision 8): the drawn illustration bundled with the app
   * (apps/mobile/assets/examples/<slot>@2x.png) and Brock's legend — one line per numbered badge,
   * rendered by the app, never baked into the image. Absent/null → the federal sample path.
   */
  illustration?: { slot: string; aspect: 'portrait' | 'landscape' } | null;
  legend?: string[];
  /**
   * The federal sample's "look for this" callouts (SBC, MSN) — sent whenever a sample exists, even
   * once it is illustrated: the app shows them (with source_line) when its build lacks the image.
   */
  callouts: string[];
  asset: { kind: 'external_pdf'; url: string; publisher: string } | null;
  source_line: string | null;
  glosses: Record<string, string>;
}

export interface IntakeHelp {
  document_type: string;
  scope: 'payer' | 'generic';
  payer_name: string | null;
  title: string | null;
  note: string | null;
  steps: string[];
  verified: boolean;
  can_email: boolean;
}

/** One screen, as the planner hands it over. `copy` is registry text keyed by slot — the app
 *  holds no intake copy. `example` / `help` are PRESENT only when there is something behind them. */
export interface IntakeScreen {
  id: string;
  kind: IntakeScreenKind;
  progress_group: IntakeProgressGroup | null;
  copy: Record<string, string>;
  data: Record<string, unknown>;
  skippable: boolean;
  example?: IntakeExample;
  help?: IntakeHelp;
}

export interface IntakeProgress {
  segments: { group: IntakeProgressGroup; filled: boolean; label: string | null }[];
  filled: number;
  total: number;
  /** "1 of 7 — nice start". Null before anything lands; never a bare "Step N of M". */
  line: string | null;
  /** Set when a segment is held by the high-water mark alone (a document was reclassified). */
  note: string | null;
  glosses: Record<string, string | null>;
  high_water: string[];
}

export interface IntakeResume {
  case_file_id: string;
  title: string | null;
  body: string | null;
  primary: string | null;
  new: string | null;
  /** States the REAL sign-in link lifetime. */
  link_expiry: string | null;
}

/**
 * The 14 canonical coverage regimes (Brock 2026-07-06). Mirrors runtime app/plan_types.py
 * PLAN_TYPES exactly — kept in sync by tests/test_plan_types_canonical.py.
 */
export const COVERAGE_REGIMES = [
  'state_regulated_commercial',
  'erisa_self_funded',
  'medicare_traditional',
  'medicare_advantage',
  'medicaid_ffs',
  'medicaid_mco',
  'dual_eligible',
  'self_pay',
  'tricare',
  'va_champva',
  'fehb_pshb',
  'nonfederal_governmental',
  'stldi',
  'excepted_coverage',
] as const;

export type CoverageRegime = (typeof COVERAGE_REGIMES)[number];

/** Typed coverage attributes (Brock 2026-07-06) — validated against the regime in the runtime. */
export interface CoverageAttributes {
  qmb_status?: boolean | null;
  ihs_prc_eligible?: boolean | null;
  grandfathered?: boolean | null;
  market_segment?: string | null;
  church_plan?: boolean | null;
  medigap?: boolean | null;
  dsnp?: boolean | null;
  governmental_fully_insured?: boolean | null;
}

export interface ConfirmationPrompt {
  field: string;
  read_value: string;
  prompt: string;
  confidence: number;
}

/** Detection metadata surfaced to the confirm screen + settings (DL-82). */
export interface RegimeDetection {
  regime: CoverageRegime | null;
  candidate: CoverageRegime | null;
  confidence: 'high' | 'medium' | 'low';
  method:
    | 'document_format'
    | 'card_branding'
    | 'member_id_pattern'
    | 'user_declared'
    | 'backfill'
    | 'ambiguous';
  evidence: string[];
  verified: boolean;
}

export interface IntakeCapturedData {
  coverage: Record<string, unknown>;
  bills_count: number;
  eobs_count: number;
  visit_context: string | null;
  /** The case's detected/confirmed coverage regime (DL-82); null until settled. */
  coverage_regime: CoverageRegime | null;
  regime_detection: RegimeDetection | null;
}

/** A stored plan-level benefit design proposed for one-tap confirmation (CO-12C). */
export interface PlanProposal {
  plan_library_id: string;
  payer: string;
  plan_name: string | null;
  plan_year: number;
  benefit_design: Record<string, unknown>;
  confidence: number;
  summary: string;
}

export interface IntakeStateResponse {
  /** Null until POST /v1/intake/start — opening the landing creates nothing. */
  case_file_id: string | null;
  intake_status: IntakeStatus;
  intake_mode: 'guided' | 'chat_first';
  /** The planner's pick: a screen id, or 'READY'. */
  current_step: string;
  /** The filled progress groups. */
  completed_steps: string[];
  screen: IntakeScreen;
  progress: IntakeProgress;
  /** intake.chrome.* — save & exit, see an example, help me find it, errors. */
  chrome: Record<string, string>;
  /** "Pick up where you left off" — only when the user returns with an unfinished case. */
  resume?: IntakeResume | null;
  captured_data: IntakeCapturedData;
  missing_items: string[];
  plan_proposal?: PlanProposal | null;
}

/** Every intake write returns the NEXT state — the client never decides where to go. */
export interface IntakeStepAck extends IntakeStateResponse {
  confirmations: ConfirmationPrompt[];
}

export interface IntakeRunResponse {
  case_file_id: string;
  status: string;
  /** The existing reveal/thread — the guided route builds no results UI. */
  next_route: string;
  conversation_id: string | null;
}

export interface IntakeCompletionSummary {
  case_file_id: string;
  intake_status: IntakeStatus;
  captured: string[];
  missing_items: string[];
  summary: string;
}
